programmer-skills/sorting-algorithms.md

```diff
@@ -0,0 +1,277 @@
+
+#####Javascript implementation of insertion sort:
+
+```javascript
+function insertionSort(test){
+	for (var i = 1; i < test.length; i++)
+	{
+		var j = i;
+		while ( j > 0 && test[j-1] > test[j])
+		{
+			var temp = test[j-1];
+			test[j-1] = test[j];
+			test[j] = temp;
+			j--;
+		}
+	}
+	return test;
+}
+```
+
+###Quicksort
+
+Quicksort is a divide and conquer algorithm. The algorithm can be borken down to two basic operations:
+ swapping items and partitioning a section of the array.
+
+
+Algorithm in steps:  
+1. Choose the pivot value (you compare all other elements against the pivot at first).    
+2. Partition: rearrange the array in such a way that it is divided into two parts. 
+All the elements smaller than the pivot element go before the pivot that is the left part. All the
+elements greater than the pivot go after the pivot element that is the right part.
+All the elements in the left and right sub-arrays must satisfy *left ≤ pivot ≤ right* inequality.  
+3. Sort both parts of the array. Recursively invoke quicksort to the rigth and left part.
+
+Partition algorithm:  
+1. Define two variables:  
+&emsp;&emsp;left: pointing at the first item in the array,  
+&emsp;&emsp;right: pointing at the last item in  the array.  
+2. While the value at array[left] is less than the pivot value, add 1 to left.
+Continue until you reach an item which is greater or equal to the pivot value.    
+3.While the value at array[right] is greater than the pivot value, substract one from right.
+Continue until you reach an item which is lesser or equal to the pivot value.    
+4.If array[left] is less than or equal to array[right], then swap the two values.  
+5. Add 1 to left,and substract one to right.  
+6. if left and right indices point to the same element (array[left]==array[right]) STOP, otherwise start again from step 1.
+
+There are many approaches to choose the pivot item.
+* select the first item in the array. If the array is already sorted, this choice has the worst performance.
+* select the middle item in the array. The latter implementation will follow this approach.
+
+Figure 5. explains the quicksort algorithm:  
+![Quicsort](http://www.tutorialspoint.com/data_structures_algorithms/images/quick_sort_partition_animation.gif)    	  
+*Fig5. is taken from  [Tutorialspoint](http://www.tutorialspoint.com/data_structures_algorithms)* 
+
+#####Javascript implementation of quicksort:
+
+```javascript
+function swap(array,index1,index2){
+	var temp = array[index1];
+	array[index1] = array[index2];
+	array[index2] = temp;
+}
+
+function partition(array,left,right){
+	var pivot = Math.floor( (left + right)/2 );
+	var i = left;
+	var j = right;
+	while ( i < j)
+	{
+		while ( array[i] < array[pivot]){
+			i++;
+		}
+		while ( array[j] > array[pivot] )
+		{
+			j--;
+		}
+		
+		if (i <= j)
+		{
+			swap(array, i, j);
+			i++;
+			j--;
+		}
+		
+	}
+	return i;
+}
+
+
+function quickSort(array,left,right){
```

Comment:
Why is an input of left and right required?  Presumably if you are sorting the whole array, left will be 0 and right will be array.length -1?

I understand that this allows you to select a portion of the array to sort, but as this isn't an aspect unique to quicksort perhaps it would be better to remove this functionality in the tutorial to make it cleaner to follow?